Plugins/Sycl/src/Seeding/LinearTransform.hpp

```diff
@@ -0,0 +1,134 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+// SYCL plugin include(s).
+#include "Acts/Plugins/Sycl/Seeding/detail/Types.hpp"
+#include "../Utilities/Arrays.hpp"
+#include "SpacePointType.hpp"
+
+// SYCL include(s).
+#include <CL/sycl.hpp>
+
+// System include(s).
+#include <cassert>
+#include <cstdint>
+
+namespace Acts::Sycl::detail {
+
+/// Functor performing a linear coordinate transformation on spacepoint pairs
+template <SpacePointType OtherSPType>
+class LinearTransform {
+  // Sanity check(s).
+  static_assert((OtherSPType == SpacePointType::Bottom) ||
+                    (OtherSPType == SpacePointType::Top),
+                "Class must be instantiated with either "
+                "Acts::Sycl::detail::SpacePointType::Bottom or "
+                "Acts::Sycl::detail::SpacePointType::Top");
+
+ public:
+  /// Constructor with all the necessary arguments
+  LinearTransform(uint32_t nMiddleSPs,
+                  const device_array<DeviceSpacePoint>& middleSPs,
+                  uint32_t nOtherSPs,
+                  const device_array<DeviceSpacePoint>& otherSPs,
+                  const device_array<uint32_t>& middleIndexLUT,
+                  const device_array<uint32_t>& otherIndexLUT, uint32_t nEdges,
+                  device_array<DeviceLinEqCircle>& resultArray)
+      : m_nMiddleSPs(nMiddleSPs),
+        m_middleSPs(middleSPs.get()),
+        m_nOtherSPs(nOtherSPs),
+        m_otherSPs(otherSPs.get()),
+        m_middleIndexLUT(middleIndexLUT.get()),
+        m_otherIndexLUT(otherIndexLUT.get()),
+        m_nEdges(nEdges),
+        m_resultArray(resultArray.get()) {}
+
+  /// Operator performing the coordinate linear transformation
+  void operator()(cl::sycl::nd_item<1> item) const {
+    // Get the index to operate on.
+    const auto idx = item.get_global_linear_id();
+    if (idx >= m_nEdges) {
+      return;
+    }
+
+    // Translate this one index into indices in the spacepoint arrays.
+    // Note that using asserts with the CUDA backend of dpc++ is not working
+    // quite correctly at the moment. :-( So these checks may need to be
+    // disabled if you need to build for an NVidia backend in Debug mode.
+    const uint32_t middleIndex = m_middleIndexLUT[idx];
+    assert(middleIndex < m_nMiddleSPs);
+    (void)m_nMiddleSPs;
+    const uint32_t otherIndex = m_otherIndexLUT[idx];
+    assert(otherIndex < m_nOtherSPs);
+    (void)m_nOtherSPs;
```

Comment:
I don't entirely understand why this is better than the check in the DupletSearch kernel
```cpp
if ((middleIndex >= m_nMiddleSPs) || (otherIndex >= m_nOtherSPs)) {
    return;
}
```

Reply:
:smile: The kernel being executed on more indices than absolutely necessary is being checked in line 57. These checks are different. These make sure that the logic in translating `idx` into a `middleIndex` and `otherIndex` seems to make sense. If those values don't fulfill these asserts, that must mean a bug in the code.

I littered the CUDA code with tons of asserts, as they helped during my code development. I wanted to add some more to the SYCL code as well, but after I noticed that Intel's compiler is having big issues with compiling these into NVidia binaries, I rather stopped. But I hope that further updates to the compiler will fix these issues. :wink:

Reply:
Hmm... I looked a bit at the code today once again... It seems that these asserts are here for a very good reason. As they are actually triggering in the code!

At first I thought that I made a mistake, but even the code from the current master, when running on Intel hardware, does one of two things:
  - crash in debug builds;
  - produce a very low agreement with the CPU code in optimised builds.

I am **very** confused by this right now. I was absolutely convinced that I tested this myself as well before, and never noticed such problems.

So now I'll need to set up one of my machines with a proper (Linux) kernel for using `gdb-oneapi`...

Reply:
It seems that my issues are coming from some bug(s) in oneAPI rather than from our code. Since I only see problems in builds made with oneAPI Beta10. With that build, I can just not get reliable results out of my integrated GPU.

```
[bash][Elrond]:build-oneapi > ./bin/ActsUnitTestSeedfinderSycl -f ../atlas_spacepoints/pu20/evt1.txt -m
14:24:40    Sycl::QueueW   INFO      Running on: Intel(R) Graphics Gen9 [0x1912]
read 33690 SP from file ../atlas_spacepoints/pu20/evt1.txt
Preparation time: 0.211781
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU       SYCL  Speedup/ Agreement
           Time (s):   0.077847   0.661106            0.117753
        Seeds found:       6665       4191           56.444107
---------------------------------------------------------------

[bash][Elrond]:build-oneapi >
```

(Notice the very low agreement between the CPU and GPU results.) But when using a private build of the [intel/llvm](https://github.com/intel/llvm) code as it was in GitHub this morning, the results look fine on both NVidia and Intel devices.

```
[bash][Elrond]:build-nvidia > ./bin/ActsUnitTestSeedfinderSycl -f ../atlas_spacepoints/pu20/evt1.txt -d "GeForce GTX 960" -m
14:28:50    Sycl::QueueW   INFO      Running on: GeForce GTX 960
read 33690 SP from file ../atlas_spacepoints/pu20/evt1.txt
Preparation time: 0.229034
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU       SYCL  Speedup/ Agreement
           Time (s):   0.076882   0.140914            0.545599
        Seeds found:       6665       6665          100.000000
---------------------------------------------------------------

[bash][Elrond]:build-nvidia >
```

```
[bash][Elrond]:build-intel > ./bin/ActsUnitTestSeedfinderSycl -f ../atlas_spacepoints/pu20/evt1.txt -m
14:24:46    Sycl::QueueW   INFO      Running on: Intel(R) Graphics Gen9 [0x1912]
read 33690 SP from file ../atlas_spacepoints/pu20/evt1.txt
Preparation time: 0.200547
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU       SYCL  Speedup/ Agreement
           Time (s):   0.077469   0.772708            0.100256
        Seeds found:       6665       6665          100.000000
---------------------------------------------------------------

[bash][Elrond]:build-intel >
```

So this PR should not be held up by this. I'll follow up with the Intel developers about the issue.